docs/migrations/support-slashed-denoms.md

```diff
@@ -0,0 +1,97 @@
+```go
+github.com/cosmos/ibc-go/v3 -> github.com/cosmos/ibc-go/v4
+```
+
+This document is necessary when chains are upgrading from a version that does not support slashed base denoms (e.g. v3.0.0) to a version that does (e.g. v3.1.0).
+
+If a chain receives a slashed denom before it upgrades to supporting it, the receive may pass however the trace information will be incorrect.
+
+E.g. If a base denom of `testcoin/testcoin/testcoin` is sent to a chain that does not support slashes in the base denom; the receive will be successful. However, the trace information stored on the receiving chain will be: `Trace: "testcoin/testcoin", BaseDenom: "testcoin"`
+
+This incorrect trace information must be corrected when the chain does upgrade to fully supporting slashed denominations.
+
+To do so, chain binaries should include a migration script that will run when the chain upgrades from not supporting slashed base denominations to supporting slashed base denominations.
+
+## Chains
+
+### Transfer
+
+The transfer module will now support slashes in base denoms, so we must iterate over current traces to check if any of them are incorrectly formed and correct the trace information.
+
+### Upgrade Proposal
+
+```go
+// Here the upgrade name is the upgrade name set by the chain
+app.UpgradeKeeper.SetUpgradeHandler("supportSlashingDenomUpgrade",
+    func(ctx sdk.Context, _ upgradetypes.Plan, fromVM module.VersionMap) (module.VersionMap, error) {
+        // list of traces that must replace the old traces in store
+        var newTraces []transfertypes.DenomTrace
+
+        transferKeeper.IterateDenomTraces(ctx,
+        func(dt transfertypes.DenomTrace) bool {
+            // check if the new way of splitting FullDenom
+            // into Trace and BaseDenom is the same as the current
+            // DenomTrace.
+            // If it isn't then store the new DenomTrace in the list of new traces.
+            newTrace := transfertypes.ParseDenomTrace(dt.GetFullDenomPath())
+
+            if !reflect.DeepEqual(newTrace, dt) {
+                append(newTraces, newTrace)
+            }
+        })
+
+        // replace the outdated traces with the new trace information
+        for _, nt := range newTraces {
+            transferKeeper.SetDenomTrace(ctx, nt)
+        }
+    }
+)
+```
+
+This is only necessary if there are DenomTraces in the store with incorrect trace information from previously received coins that had a slash in the base denom. However, it is recommended that any chain upgrading to support slashed denominations runs this code for safety.
+
+### Genesis Migration
+
+If the chain chooses to add support for slashes in base denoms via genesis export, then the trace information must be corrected during genesis migration.
+
+The migration code required may look like:
+
+```go
+func MigrateGenesis(appState genutiltypes.AppMap, clientCtx client.Context, genDoc tmtypes.GenesisDoc) (genutiltypes.AppMap, error) {
```

Comment:
This code also needed some adjustments to compile:

```go
func MigrateGenesis(appState genutiltypes.AppMap, clientCtx client.Context, genDoc *tmtypes.GenesisDoc) (genutiltypes.AppMap, error) {
	if appState[ibctransfertypes.ModuleName] != nil {
		transferGenState := &ibctransfertypes.GenesisState{}
		clientCtx.Codec.MustUnmarshalJSON(appState[ibctransfertypes.ModuleName], transferGenState)

		substituteTraces := make([]ibctransfertypes.DenomTrace, len(transferGenState.DenomTraces))
		for i, dt := range transferGenState.DenomTraces {
			// replace all previous traces with the latest trace
			// note most traces will have same value
			newTrace := ibctransfertypes.ParseDenomTrace(dt.GetFullDenomPath())

			substituteTraces[i] = newTrace
		}

		transferGenState.DenomTraces = substituteTraces

		// delete old genesis state
		delete(appState, ibctransfertypes.ModuleName)

		// set new ibc transfer genesis state
		appState[ibctransfertypes.ModuleName] = clientCtx.Codec.MustMarshalJSON(transferGenState)
	}

	return appState, nil
}
```

Reply:
I haven't been able to test this migration because the [app state export panics](https://github.com/cosmos/cosmos-sdk/issues/12227).

Reply:
I did a trick (I skipped the app state export, but instead I manually modified the genesis.json file to add malformed denom traces) and I was able to test this migration functionality. It also works as expected.